### ISSUE TITLE
WANTED Poster Greets You

### ISSUE THREAD

**Issue by PatSmuk** [classicdb/database#142]
The WANTED poster in Thelsamar greets you. Enough said.

**Comment by cala** [comment]
I think almost all gameobjects with a gossip menu have this default behaviour when the condition to display the intended gossip is not met (quest accepted or available, mainly).

Thus, I assumed it is a core limitation rather than something missing/wrong DB side.

**Comment by Schmoozerd** [comment]
What is supposed to happen? ;)

Probably a core bug..

**Comment by cala** [comment]
It is not supposed to happen. :P 

As this is a minor issue, I did not report it or look at it yet.

**Comment by PatSmuk** [comment]
Well you shouldn't be able to use the object at all once you have the quest. I'll look further into it sometime later.

**Comment by cala** [comment]
After peering a bit in core code I _think_ the logic in the following code does not work as intended in `gameobject.cpp`. It looks like the GO activates for the player even if he/she has already completed the quest (not sure if it activates if player is not yet eligible for quest). 

``` c++
bool GameObject::ActivateToQuest(Player* pTarget) const
{
    // if GO is ReqCreatureOrGoN for quest
    if (pTarget->HasQuestForGO(GetEntry()))
        return true;

    if (!sObjectMgr.IsGameObjectForQuests(GetEntry()))
        return false;

    switch (GetGoType())
    {
        case GAMEOBJECT_TYPE_QUESTGIVER:
        {
            // Not fully clear when GO's can activate/deactivate
            // For cases where GO has additional (except quest itself),
            // these conditions are not sufficient/will fail.
            // Never expect flags|4 for these GO's? (NF-note: It doesn't appear it's expected)

            QuestRelationsMapBounds bounds = sObjectMgr.GetGOQuestRelationsMapBounds(GetEntry());

            for (QuestRelationsMap::const_iterator itr = bounds.first; itr != bounds.second; ++itr)
            {
                const Quest* qInfo = sObjectMgr.GetQuestTemplate(itr->second);

                if (pTarget->CanTakeQuest(qInfo, false))
                    return true;
            }

            bounds = sObjectMgr.GetGOQuestInvolvedRelationsMapBounds(GetEntry());

            for (QuestRelationsMap::const_iterator itr = bounds.first; itr != bounds.second; ++itr)
            {
            // THE LINE UNDER SHOULD PROBABLY BE IMPROVED
                if ((pTarget->GetQuestStatus(itr->second) == QUEST_STATUS_INCOMPLETE || pTarget->GetQuestStatus(itr->second) == QUEST_STATUS_COMPLETE)
                        && !pTarget->GetQuestRewardStatus(itr->second))
                    return true;
            }

            break;
        }
```

Thus, the second code tries to send a gossip menu to the player and as most of the quest-giver GOs don't have one, the GO displays the default "Greetings, $name" gossip.

``` c++
void GameObject::Use(Unit* user)
    switch (GetGoType())
    {
        case GAMEOBJECT_TYPE_QUESTGIVER:                    // 2
        {
            if (user->GetTypeId() != TYPEID_PLAYER)
                return;

            Player* player = (Player*)user;

            if (!sScriptMgr.OnGossipHello(player, this))
            {
                player->PrepareGossipMenu(this, GetGOInfo()->questgiver.gossipID);
                player->SendPreparedGossip(this);
            }

            return;
        }
```

I think the best option would be to prevent the GO to activate when it should'nt and let the gossip code untouched. What do you think? (just trying here to _slowly_ understand core code step by step)

**Comment by Schmoozerd** [comment]
Why do you suggest to change the rewarded-relations?
In this case, why does this part return true even if it should not?

**Comment by cala** [comment]
This suggestion comes from a wrong assumption of mine: all gameobjects with GAMEOBJECT_TYPE_QUESTGIVER _without a gossip assigned in DB_ display default gossip and are usable whatever the player status regarding the quest is. Thus, I assumed there was an error in the code activating the GO based on quest rewarded-relations making it always return true.

However after testing this further, the issue is slightly different. Sorry about that.

How it is:
- GOs with GAMEOBJECT_TYPE_QUESTGIVER are always usable.
- If they don't have a gossip assigned in DB: "Greetings, $name" default gossip is displayed.
- If they have a gossip assigned in DB: correct gossip is displayed. But, if gossip does not exist (missing in DB): "Greetings, $name" default gossip is displayed and no error is put in DBErrors.log 

How it should be:
- GOs with GAMEOBJECT_TYPE_QUESTGIVER are always usable IF gossip is assigned
- GOs with GAMEOBJECT_TYPE_QUESTGIVER without gossip are only usable if quest reward-relations allow player to use it
- If gossip is assigned but does not exist (missing in DB): "Greetings, $name" default gossip is displayed and an error should be put in DBErrors.log stating that GO 1234 (type 2) is using gossip menu 5678 but that table gossip_menu don't have entry for that menu

So, I think I should look somewhere else in the code to understand what is going on and how to fix this. :)

**Comment by Gerhood** [comment]
Hi, here's how did it on trinity in GameObject::BuildValuesUpdate.
Originally:

```
...
switch (GetGoType())
{
case GAMEOBJECT_TYPE_QUESTGIVER:
    if (ActivateToQuest(target))
        dynFlags |= GO_DYNFLAG_LO_ACTIVATE;
    break;
...
```

Quest-GOs seem to be usable regardless of the GO_DYNFLAG_LO_ACTIVATE flag.
But using GO_DYNFLAG_LO_NO_INTERACT works:

```
...
switch (GetGoType())
{
case GAMEOBJECT_TYPE_QUESTGIVER:
    if (!ActivateToQuest(target) && !GetGOInfo()->questgiver.gossipID)
        dynFlags |= GO_DYNFLAG_LO_NO_INTERACT; 
    break;
...
```

**Comment by cala** [comment]
That's a very nice feedback, thank you. I will test that. :+1:

**Comment by lduguid** [comment]
I was motivated to take a look at this after @Phatcat mentioned a particular issue in [Gadgetzan #860](https://github.com/classicdb/database/issues/860) and @Tobschinski linked back to this issue.

As mentioned by @Gerhood above, I attempted to appropriately set **GO_DYNFLAG_LO_NO_INTERACT** in **GameObject::BuildValuesUpdate** but this did not appear to have any effect on suppressing the greeting message when actually interacting with the quest object. I reviewed the current Trinity core source, as he mentioned that was a viable fix for them but it seems time has marched forward and GameObject::BuildValuesUpdate(...) is quite different now, I could not see **GO_DYNFLAG_LO_NO_INTERACT** being referenced anywhere...

Anyway, reading over @Cala s initial investigation and following the stack trace when ddebugging I moved this logic to the bottom of **SpellHandler.cpp, void WorldSession::HandleGameObjectUseOpcode(WorldPacket& recv_data)**

```
// Never expect this opcode for non intractable GO's
    if (obj->HasFlag(GAMEOBJECT_FLAGS, GO_FLAG_NO_INTERACT))
    {
        sLog.outError("HandleGameObjectUseOpcode: CMSG_GAMEOBJ_USE for GameObject (Entry %u) with non intractable flag (Flags %u), didn't expect this to happen.", obj->GetEntry(), obj->GetUInt32Value(GAMEOBJECT_FLAGS));
        return;
    }

// from here --->
// GO is quest object and cannot activate to quest and default greeting message
if (obj->GetGoType() == GAMEOBJECT_TYPE_QUESTGIVER && !obj->ActivateToQuest(_player) && !obj->GetGOInfo()->GetGossipMenuId())
        return;
/// <--- to here 

obj->Use(_player);
```

and it appears to have the desired outcome. I suspect that leaving this kind of state 'check' until HandleGameObjectUseOpcode is probably frowned upon but I have never done any core modifications before and do not know if it appropriate or not? If it is worth someone here actually taking a second look at it, I can create a pull request.

**Comment by Phatcat** [comment]
@lduguid Always feel free to do a PR. You can always trust on me looking it over if you tag me in it but I'm not sure I may be right person to review this change, but a PR makes it a lot easier for the right person to review and test this.

**Comment by Gerhood** [comment]
@lduguid I just had a look at the Mangos source and logic is a bit different than with Trinity. The disabling should be done right here: https://github.com/cmangos/mangos-classic/blob/22b0a55d3384cf8905f76260a8f1fd3aab7b130f/src/game/Object.cpp#L520, not in the switch.
I am not a TC developer and I have not yet proposed this change to them thats why you cant find it in their repository.

**Comment by lduguid** [comment]
@Gerhood. Thank you for the help, as you can tell, I am out of my comfort zone on this one.

Concerning your suggestion, this is indeed what and where I made my initial modification after reading your original proposal. It looked something like this;

```
 else if (isType(TYPEMASK_GAMEOBJECT))                   // gameobject case
    {
        for (uint16 index = 0; index < m_valuesCount; ++index)
        {
            if (updateMask->GetBit(index))
            {
                // send in current format (float as float, uint32 as uint32)
                if (index == GAMEOBJECT_DYN_FLAGS)
                {
                    GameObject const* gameObject = static_cast<GameObject const*>(this);
                    if (IsActivateToQuest)
                    {
                        //GameObject const* gameObject = static_cast<GameObject const*>(this);
                        switch (gameObject->GetGoType())
                        {
                            case GAMEOBJECT_TYPE_QUESTGIVER:
                            case GAMEOBJECT_TYPE_CHEST:
                                if (gameObject->getLootState() == GO_READY || gameObject->getLootState() == GO_ACTIVATED)
                                    *data << uint16(GO_DYNFLAG_LO_ACTIVATE | GO_DYNFLAG_LO_SPARKLE);
                                else
                                    *data << uint16(0);
                                *data << uint16(0);
                                break;
                            case GAMEOBJECT_TYPE_GENERIC:
                            case GAMEOBJECT_TYPE_SPELL_FOCUS:
                            case GAMEOBJECT_TYPE_GOOBER:
                                *data << uint16(GO_DYNFLAG_LO_ACTIVATE);
                                *data << uint16(0);
                                break;
                            default:
                                *data << uint32(0);         // unknown, not happen.
                                break;
                        }
                    }
                    else {
                        // from here -->
                        if (gameObject->GetGoType == GAMEOBJECT_TYPE_QUESTGIVER &&
                            !gameObject->ActivateToQuest(target) && !gameObject->GetGOInfo()->GetGossipMenuId()) {
                             updateMask->SetBit(GAMEOBJECT_DYN_FLAGS);
                            *data << uint16(GO_DYNFLAG_LO_NO_INTERACT);
                            *data << uint16(0);
                        }
                        else
                        // <-- to here
                            *data << uint32(0);                 // disable quest object
                    }
                }
                else
                    *data << m_uint32Values[index];         // other cases
            }
        }
```

 I tried it again this morning and it appeared to be working on most appropriate quest game objects, like the wanted posters and the conspicuous tomestone! I should have been more diligent in testing them **all** post my initial change. I believe where I got it wrong, was that after I made this particular change I tested it on the old lion statue, which has a different player interaction. i.e it shows a progress bar (like opening a chest) with the user interacts with it and still shows the default greeting message regardless of this change.  This may be the desired behavior for quests GO's that are like this _shrug_? 

 Bottom line is that your suggested change goes along way in removing the default greeting message from the bulk of quest GO's that should not display it when the player is not appropriate for that particular quest. And I think that the general consensus is that this is the better place to make this change rather than in other code locations.

Thank you again for your help. I will drop my original pull request later tonight and resubmit one that looks similar to the code snippet above (unless someone sees a problem with it ;) after I am done with work tonight.
